UnicodeDecodeError (INTERNALERROR) when doctests contain Unicode

Originally reported by: **Jason R. Coombs (BitBucket: [jaraco](http://bitbucket.org/jaraco), GitHub: [jaraco](http://github.com/jaraco))**

----------------------------------------

Consider this test:

```
def foo():
    """
    >>> name = 'с'
    'anything'
    """
```

Note that isn't the letter 'c' but instead Cyrillic 's'.

Save it as mod.py, then invoke `pytest --doctest-modules`. It will fail with this output:

```
> py.test --doctest-modules
﻿============================= test session starts =============================

platform win32 -- Python 3.4.2 -- py-1.4.26 -- pytest-2.6.4
collected 1 items

mod.py
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\main.py", line 84, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\main.py", line 122, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 413, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\main.py", line 142, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 413, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\runner.py", line 65, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\runner.py", line 75, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\runner.py", line 111, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)

INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\main.py", line 167, in call_matching_hooks
INTERNALERROR>     return hookmethod.pcall(plugins, **kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 417, in pcall
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 424, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\core.py", line 315, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\runner.py", line 214, in pytest_runtest_makereport
INTERNALERROR>     longrepr = item.repr_failure(excinfo)
INTERNALERROR>   File "c:\python\lib\site-packages\pytest-2.6.4-py3.4.egg\_pytest\doctest.py", line 62, in repr_failure
INTERNALERROR>     filelines = py.path.local(filename).readlines(cr=0)
INTERNALERROR>   File "c:\python\lib\site-packages\py-1.4.26-py3.4.egg\py\_path\common.py", line 141, in readlines
INTERNALERROR>     content = self.read('rU')
INTERNALERROR>   File "c:\python\lib\site-packages\py-1.4.26-py3.4.egg\py\_path\common.py", line 135, in read
INTERNALERROR>     return f.read()
INTERNALERROR>   File "c:\python\lib\encodings\cp1252.py", line 23, in decode
INTERNALERROR>     return codecs.charmap_decode(input,self.errors,decoding_table)[0]
INTERNALERROR> UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 36: character maps to <undefined>

==============================  in 0.30 seconds ===============================
```

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/628

On OS X, the test fails as expected:

```
$ py.test test.py --doctest-modules
====================================== test session starts =======================================
platform darwin -- Python 3.4.3 -- py-1.4.30 -- pytest-2.7.2
rootdir: /Users/jaraco/Dropbox/code/public/pytest, inifile: tox.ini
collected 1 items 

test.py F

============================================ FAILURES ============================================
_______________________________________ [doctest] test.foo _______________________________________
002     """
003     >>> name = 'с'
Expected:
    'anything'
Got nothing

/Users/jaraco/Dropbox/code/public/pytest/issue628/test.py:3: DocTestFailure
==================================== 1 failed in 0.08 seconds ====================================
```

On that system, getfilesystemencoding returns 'utf-8'.

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -19,6 +19,9 @@
   properly displayed.
   Thanks Ionel Maries Cristian for the report and Bruno Oliveira for the PR.
 
+- fix #628: fixed internal UnicodeDecodeError when doctests contain unicode.
+  Thanks Jason R. Coombs for the report and Bruno Oliveira for the PR.
+
 
 2.8.5
 -----

--- a/_pytest/doctest.py
+++ b/_pytest/doctest.py
@@ -81,15 +81,15 @@ class DoctestItem(pytest.Item):
             reprlocation = ReprFileLocation(filename, lineno, message)
             checker = _get_unicode_checker()
             REPORT_UDIFF = doctest.REPORT_UDIFF
-            filelines = py.path.local(filename).readlines(cr=0)
-            lines = []
             if lineno is not None:
-                i = max(test.lineno, max(0, lineno - 10)) # XXX?
-                for line in filelines[i:lineno]:
-                    lines.append("%03d %s" % (i+1, line))
-                    i += 1
+                lines = doctestfailure.test.docstring.splitlines(False)
+                # add line numbers to the left of the error message
+                lines = ["%03d %s" % (i + test.lineno + 1, x)
+                         for (i, x) in enumerate(lines)]
+                # trim docstring error lines to 10
+                lines = lines[example.lineno - 9:example.lineno + 1]
             else:
-                lines.append('EXAMPLE LOCATION UNKNOWN, not showing all tests of that example')
+                lines = ['EXAMPLE LOCATION UNKNOWN, not showing all tests of that example']
                 indent = '>>>'
                 for line in example.source.splitlines():
                     lines.append('??? %s %s' % (indent, line))

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -1,3 +1,4 @@
+# encoding: utf-8
 import sys
 from _pytest.doctest import DoctestItem, DoctestModule, DoctestTextfile
 import py
@@ -108,6 +109,46 @@ class TestDoctests:
             "*>>> 0 / i*",
             "*UNEXPECTED*ZeroDivision*",
         ])
+
+    def test_docstring_context_around_error(self, testdir):
+        """Test that we show some context before the actual line of a failing
+        doctest.
+        """
+        testdir.makepyfile('''
+            def foo():
+                """
+                text-line-1
+                text-line-2
+                text-line-3
+                text-line-4
+                text-line-5
+                text-line-6
+                text-line-7
+                text-line-8
+                text-line-9
+                text-line-10
+                text-line-11
+                >>> 1 + 1
+                3
+
+                text-line-after
+                """
+        ''')
+        result = testdir.runpytest('--doctest-modules')
+        result.stdout.fnmatch_lines([
+            '*docstring_context_around_error*',
+            '005*text-line-3',
+            '006*text-line-4',
+            '013*text-line-11',
+            '014*>>> 1 + 1',
+            'Expected:',
+            '    3',
+            'Got:',
+            '    2',
+        ])
+        # lines below should be trimmed out
+        assert 'text-line-2' not in result.stdout.str()
+        assert 'text-line-after' not in result.stdout.str()
 
     def test_doctest_linedata_missing(self, testdir):
         testdir.tmpdir.join('hello.py').write(py.code.Source("""
@@ -338,6 +379,23 @@ class TestDoctests:
         """)
         reprec = testdir.inline_run(p, "--doctest-glob=x*.txt")
         reprec.assertoutcome(failed=1, passed=0)
+
+    def test_contains_unicode(self, testdir):
+        """Fix internal error with docstrings containing non-ascii characters.
+        """
+        testdir.makepyfile(u'''
+            # encoding: utf-8
+            def foo():
+                """
+                >>> name = 'с' # not letter 'c' but instead Cyrillic 's'.
+                'anything'
+                """
+        ''')
+        result = testdir.runpytest('--doctest-modules')
+        result.stdout.fnmatch_lines([
+            'Got nothing',
+            '* 1 failed in*',
+        ])
 
     def test_ignore_import_errors_on_doctest(self, testdir):
         p = testdir.makepyfile("""

